src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -53,135 +53,164 @@ namespace impl {
- *       detail, and, normally, should not be used by customer apps
  */
-template <bool Serialize, typename TeamMember, typename Lambda, typename ValueType>
+template <bool Serialize, typename ValueType, typename TeamMember, typename Lambda>
 static KOKKOS_INLINE_FUNCTION
 void parallel_reduce (const TeamMember& team,
-                      const int& begin, // pack index
-                      const int& end, // pack index
+                      const int& begin,
+                      const int& end,
                       const Lambda& lambda,
                       ValueType& result)
 {
   if (Serialize) {
     // We want to get C++ on GPU to match F90 on CPU. Thus, need to
     // serialize parallel reductions.
 
-    // All threads init result.
-    // NOTE: we *need* an automatic temporary, since we do not know
-    //       where 'result' comes from. If result itself is an automatic
-    //       variable, using it would be fine (only one vector lane would
-    //       actually have a nonzero value after the single). But if
-    //       result is taken from a view, then all vector lanes would
-    //       see the updated value before the vector_reduce call,
-    //       which will cause the final answer to be multiplied by the
-    //       size of the warp.
-    auto local_tmp = result;
-
-    Kokkos::single(Kokkos::PerThread(team),[&] {
-        for (int k=begin; k<end; ++k) {
-          lambda(k, local_tmp);
-        }
-      });
-
-#ifdef EKAT_ENABLE_GPU
-    // Broadcast result to all threads by doing sum of one thread's
-    // non-0 value and the rest of the 0s.
-    team.vector_reduce(Kokkos::Sum<ValueType>(local_tmp));
-#endif
-
-   result = local_tmp;
+    // Note the version of single(..) that takes the result as last arg is handy on Cuda,
+    // since it also implements a shuffle to ensure all threads in a warp have the same
+    // result as the one in vector lane 0.
+    Kokkos::single(Kokkos::PerThread(team),[&] (ValueType& result) {
+      for (int k=begin; k<end; ++k) {
+        lambda(k, result);
+      }
+    },result);
   } else {
-    const ValueType initial(result);
-    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
-    result += initial;
+    // Kokkos::parallel_reduce inits var to 0, so we need to perform reduction on a temp,
+    // then add to result.
+    ValueType temp;
+    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, temp);
+
+    // Only one vector lane should update the result
+    Kokkos::single(Kokkos::PerThread(team),[&](ValueType& r){
+      r += temp;
+    },result);
   }
 }
 
+template <bool Serialize, typename ValueType, typename TeamMember, typename Lambda>
+static KOKKOS_INLINE_FUNCTION
+ValueType parallel_reduce (const TeamMember& team,
+                           const int& begin,
+                           const int& end,
+                           const Lambda& lambda)
+{
+  ValueType result;
```

Comment:
Note that the way you've chosen to impl `impl::parallel_reduce` above (now consistent in both cases of `Serialize`) means you _do_ need to 0-init `result` here.

As an aside, I would have chosen to make `impl::parallel_reduce` return a value, thereby making this code as simple as possible.